ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/AstSanitizer.java

```diff
@@ -0,0 +1,283 @@
+    final DataSourceExtractor dataSourceExtractor = new DataSourceExtractor(metaStore);
+    dataSourceExtractor.extractDataSources(node);
+
+    final RewriterPlugin rewriterPlugin = new RewriterPlugin(metaStore, dataSourceExtractor);
+
+    final ExpressionRewriterPlugin expressionRewriterPlugin =
+        new ExpressionRewriterPlugin(metaStore, dataSourceExtractor);
+    final BiFunction<Expression, Void, Expression> expressionRewriter =
+        (e, v) -> ExpressionTreeRewriter.rewriteWith(expressionRewriterPlugin::process, e, v);
+
+    return new StatementRewriter<>(expressionRewriter, rewriterPlugin::process)
+        .rewrite(node, null);
+  }
+
+  private static final class RewriterPlugin extends
+      AstVisitor<Optional<AstNode>, StatementRewriter.Context<Void>> {
+    final MetaStore metaStore;
+    final DataSourceExtractor dataSourceExtractor;
+
+    private int selectItemIndex = 0;
+
+    RewriterPlugin(final MetaStore metaStore, final DataSourceExtractor dataSourceExtractor) {
+      super(Optional.empty());
+      this.metaStore = Objects.requireNonNull(metaStore, "metaStore");
+      this.dataSourceExtractor
+          = Objects.requireNonNull(dataSourceExtractor, "dataSourceExtractor");
+    }
+
+    @Override
+    protected Optional<AstNode> visitCreateStreamAsSelect(
+        final CreateStreamAsSelect node,
+        final StatementRewriter.Context<Void> ctx
+    ) {
+      return Optional.of(
+          new CreateStreamAsSelect(
+              node.getLocation(),
+              node.getName(),
+              (Query) ctx.process(node.getQuery()),
+              node.isNotExists(),
+              node.getProperties(),
+              node.getPartitionByColumn()
+          )
+      );
+    }
+
+    @Override
+    protected Optional<AstNode> visitInsertInto(
+        final InsertInto node,
+        final StatementRewriter.Context<Void> ctx
+    ) {
+      final DataSource<?> target = getSource(
+          node.getTarget(),
+          node.getLocation().map(
+              l -> new NodeLocation(
+                  l.getLineNumber(),
+                  l.getColumnNumber() +  "INSERT INTO".length()
+              )
+          )
+      );
+      if (target.getDataSourceType() != DataSourceType.KSTREAM) {
+        throw new KsqlException(
+            "INSERT INTO can only be used to insert into a stream. "
+                + target.getName().toString(FormatOptions.noEscape()) + " is a table.");
+      }
+      return Optional.of(
+          new InsertInto(
+              node.getLocation(),
+              node.getTarget(),
+              (Query) ctx.process(node.getQuery()),
+              node.getPartitionByColumn()
+          )
+      );
+    }
+
+    @Override
+    protected Optional<AstNode> visitAllColumns(
+        final AllColumns allColumns,
+        final StatementRewriter.Context<Void> ctx
+    ) {
+      allColumns.getSource().ifPresent(
+          src -> throwOnUnknownNameOrAlias(dataSourceExtractor, src)
```

Comment:
This is checking that the source on the all columns is a source used in the `FROM` clause, e.g. given `SELECT FOO.*, ..`it's checking that `FOO` is a known source, eg. `FROM FOO`.

Is there any reason that this check can't remain in the AST builder?  It would seem to me that it can if the `DataSourceExtractor` does not validate that sources are known to the engine, (which is shouldn't be doing).

i.e. what I'm suggesting is that we move this check back into the AstBuilder, as it's just normal validation, and also change `DataSourceExtractor` so that it knows nothing of the metastore, i.e. it won't check its a source the engine _knows_ about.  Such a check should not be performed when _parsing_ the statement, but when _executing_ the statement.

Reply:
good call. done.